app/models/search_vector.rb

```diff
@@ -1,8 +1,76 @@
 class SearchVector < ActiveRecord::Base
+
+  WEIGHT_VALUES = [
+    0.03, # D
+    0.1,  # C
+    0.3,  # B
+    1.0   # A
+  ]
+
+  DISCUSSION_FIELD_WEIGHTS = {
+    'discussions.title'        => :A,
+    'motion_names'             => :B,
+    'discussions.description'  => :C,
+    'motion_descriptions'      => :C,
+    'comment_bodies'           => :D
+  }
+
   belongs_to :discussion
   self.table_name = 'discussion_search_vectors'
 
-  def self.execute_search_query(*args)
-    connection.execute sanitize_sql_array args
+  class << self
+    def index!(discussion_ids)
+      discussion_ids = Array(discussion_ids).map(&:to_i)
+      where(discussion_id: discussion_ids).delete_all
+      discussion_ids.each do |discussion_id|
+        connection.execute index_thread_sql(discussion_id)
+        yield if block_given?
+      end
+    end
+    handle_asynchronously :index!
+
+  end
+
+  def self.index_everything!
+    index_without_delay! Discussion.pluck(:id)
+  end
+
+  def self.index_thread_sql(discussion_id)
+    "INSERT INTO discussion_search_vectors (discussion_id, search_vector)
+     SELECT      id, #{discussion_field_weights}
+     FROM        discussions
+     LEFT JOIN (
+       SELECT string_agg(name, ',')                     AS motion_names,
+              LEFT(string_agg(description, ','), 10000) AS motion_descriptions
+       FROM   motions
+       WHERE  discussion_id = #{discussion_id}) motions ON #{discussion_id} = id
+     LEFT JOIN (
+       SELECT LEFT(string_agg(body, ','), 200000)       AS comment_bodies
+       FROM   comments
+       WHERE  discussion_id = #{discussion_id}) comments ON #{discussion_id} = id
+     WHERE    id = #{discussion_id}"
   end
+
+  def self.discussion_field_weights
+    DISCUSSION_FIELD_WEIGHTS.map { |field, weight| "setweight(to_tsvector(coalesce(#{field}, '')), '#{weight}')" }.join ' || '
+  end
+
+  scope :search_for, ->(query, user, opts = {}) do
+    Queries::VisibleDiscussions.apply_privacy_sql(
```

Comment:
The great thing about having these as scopes is that we can really easily apply weirdo filters to them. (Although we have to do it at the SearchVector level, not the Discussion level)

So, something like
```
SearchVector.search_for('bunnies', current_user).where(group: )
```

will search in a single group. This works because we're joining the discussion and group fields on the search vector query, so we have access to loads of info about the things we're searching.
(We might have to do `where('groups.id = ?', group.id)`, but who cares that still rocks)

This allows for things like per-group queries, timeframing, and searching a single field
(Search Discussion Titles Only = `.where('discussions.description ilike %#{query}%')`